File: leetcode/410. Split Array Largest Sum.rs
use std::cmp;

impl Solution {
    pub fn split_array(nums: Vec<i32>, m: i32) -> i32 {
        let m = m as usize;
        let mut dp: Vec<Vec<i32>> = vec![vec![i32::MAX; m as usize + 1]; nums.len()+1];
        
        let mut sum = 0;
        for i in (1..nums.len()+1) {
            sum += nums[i-1];
            dp[i][1] = sum;
        }
        
        for i in (1..nums.len()+1) {
            for j in (2..cmp::min(m+1, i+1)) {
                let mut sum = 0;
                for k in (j..i+1).rev() {
                    sum += nums[k-1];
                    dp[i][j] = cmp::min(dp[i][j], cmp::max(dp[k-1][j-1], sum));
                }
            }
        }
        
        dp[nums.len()][m]
    }
}

File: leetcode/410. Split Array Largest Sum.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        def can_split(target):
            cnt = 0
            curr = 0
            for i in range(len(nums)):
                if curr + nums[i] > target:
                    curr = nums[i]
                    cnt += 1
                    if cnt == k:
                        return False
                else:
                    curr += nums[i]
            return True
        
        l = max(nums)
        r = sum(nums)
        while l < r:
            mid = (l+r)//2
            if can_split(mid):
                r = mid
            else:
                l = mid + 1
        return l
